app/src/main/java/fr/neamar/kiss/searcher/Searcher.java

```diff
@@ -2,52 +2,180 @@
+		this.handler = new Handler( Looper.getMainLooper() );
+	}
 
-        if (pojos != null) {
-            for (int i = pojos.size() - 1; i >= 0; i--) {
-                results.add(Result.fromPojo(activity, pojos.get(i)));
-            }
+	protected int getMaxResultCount()
+	{
+		return DEFAULT_MAX_RESULTS;
+	}
 
-            activity.adapter.addAll(results);
-        }
-        activity.resetTask();
+	/**
+	 * This is called from the background thread by the providers
+	 */
+	public boolean addResult( Pojo... pojos )
+	{
+		if( isCancelled() )
+			return false;
 
-        long time = System.currentTimeMillis() - start;
-        Log.v("Timing", "Time to run query `" + query + "` to completion: " + time + "ms");
-    }
+		MainActivity activity = activityWeakReference.get();
+		if( activity == null )
+			return false;
+
+		this.processedPojos.addAll( Arrays.asList( pojos ) );
+		int maxResults = getMaxResultCount();
+		while( this.processedPojos.size() > maxResults )
+			this.processedPojos.poll();
+
+		int counter = this.refreshTask.runCounter;
+		// if timer passed, refresh list
+		if( this.refreshCounter < counter )
+		{
+			this.refreshCounter = counter;
+
+			// clone the PriorityQueue so we can extract the pojos in sorted order and still keep them in the original list
+			PriorityQueue<Pojo> queue   = new PriorityQueue<>( this.processedPojos );
+			Collection<Result>  results = new ArrayList<>( queue.size() );
+			while( queue.peek() != null )
+			{
+				results.add( Result.fromPojo( activity, queue.poll() ) );
+			}
+			// make the adapter update from the main thread
+			publishProgress( results.toArray( new Result[0] ) );
+		}
+
+		return true;
+	}
+
+	@CallSuper
+	@Override
+	protected void onPreExecute()
+	{
+		super.onPreExecute();
+		start = System.currentTimeMillis();
+
+		MainActivity activity = activityWeakReference.get();
+		if( activity == null )
+			return;
+
+		this.handler.postDelayed( this.refreshTask, DEFAULT_REFRESH_TIMER );
+		activity.displayLoader( true );
+	}
+
+	@Override
+	protected void onProgressUpdate( Result... results )
+	{
+		MainActivity activity = activityWeakReference.get();
+		if( activity == null )
+			return;
+
+		activity.beforeChange();
+
+		activity.adapter.clear();
+		activity.adapter.addAll( results );
+
+		activity.afterChange();
+	}
+
+	@Override
+	protected void onPostExecute( Void param )
+	{
+		this.handler.removeCallbacks( this.refreshTask );
+
+		MainActivity activity = activityWeakReference.get();
+		if( activity == null )
+			return;
+
+		activity.displayLoader( false );
+
+		if( this.processedPojos.isEmpty() )
+		{
+			activity.adapter.clear();
+		}
+		else
+		{
+			PriorityQueue<Pojo> queue   = this.processedPojos;
+			Collection<Result>  results = new ArrayList<>( queue.size() );
+			while( queue.peek() != null )
+			{
+				results.add( Result.fromPojo( activity, queue.poll() ) );
+			}
+			activity.beforeChange();
```

Comment:
all code between beforeChange() and afterChange() can probably be into a function and reused here and by onProgressUpdate